// napi/lib/comment.v2.js
var util = require('util');
var API = require('./base.js');
var fns = require('./functions')

var CommentAPI = function(appID, host, category) {
    API.call(this, appID, host);
    this._version = '2';
    this._missing('category', category);
    this._category = category;
};

util.inherits(CommentAPI, API);

function getUrl () {
    this._check('_host', '_appID', '_version');
    return [this._host, this._version, this._appID].concat([].slice.call(arguments)).join('/');
};

/**
 * 获取评论
 * @param thread {String}
 * @returns {Promise}
 */
CommentAPI.prototype.list = function(thread) {
    this._missing('thread', thread);
    return this.request({
        method: 'GET',
        url: getUrl.call(this, 'comments', this._category, thread)
    });
};

/**
 * 发表评论
 * @param thread {String}
 * @returns {Promise}
 */
CommentAPI.prototype.post = function(thread) {
    this._missing('thread', thread);
    return this.request({
        method: 'POST',
        url: getUrl.call(this, 'comments', this._category, thread)
    });
};


/**
 * 回复评论
 * @param thread {String}
 * @param commentID {String}
 * @returns {Promise}
 */
CommentAPI.prototype.reply = function(thread, commentID) {
    this._missing('thread', thread);
    this._missing('commentID', commentID);
    return this.request({
        method: 'POST',
        url: getUrl.call(this, 'comments', this._category, thread, commentID, 'reply')
    });
};

/**
 * 赞评论
 * @param thread {String}
 * @param commentID {String}
 * @returns {Promise}
 */
CommentAPI.prototype.like = function(thread, commentID) {
    this._missing('thread', thread);
    this._missing('commentID', commentID);
    return this.request({
        method: 'POST',
        url: getUrl.call(this, 'comments', this._category, thread, commentID, 'like')
    });
};

module.exports = CommentAPI;
