// napi/index.js
var fns = require('./lib/functions');
var BaseAPI = require('./lib/base.api.js');
var CommentAPI = require('./lib/comment.v2');
var CommentAPIV3 = require('./lib/comment.v3');
var ObjectAPI = require('./lib/object.v3');
var Promise = require('promise');

var NAPI = module.exports = function (appID, host, options) {
    var _conf = {
        versions: {
            object: '3', // default 3
            comment: '2' // default 2
        }
    }

    if (!appID) {
        throw new Error('missing app id');
    }
    host = host || NAPI.DEFAULT_HOST;

    if (options && options.versions) {
        extend(_conf.versions, options.versions || {})
    }

    return {
        comment: function (classID) {
            if (_conf.versions.comment == '3') {
                return new CommentAPIV3(appID, host, classID)
            } else {
                return new CommentAPI(appID, host, classID)
            }
        },
        base: function () {
            return new BaseAPI(appID, host);
        },
        object: function (classID) {
            return new ObjectAPI(appID, host, classID);
        },
        /**
         * Promise.all的包装
         * instance method
         * @returns {Promise}
         */
        all: allPromise
    }
}
/**
 *  Global method
 */
NAPI.all = allPromise

function allPromise () {
    return Promise.all.apply(Promise, arguments);
}

function extend(obj) {
    var source, prop;
    for (var i = 1, length = arguments.length; i < length; i++) {
        source = arguments[i];
        for (prop in source) {
            obj[prop] = source[prop];
        }
    }
    return obj;
}
NAPI.DEFAULT_HOST = 'http://napi.uc.cn';
